components/raftstore/src/store/peer.rs

```diff
@@ -1865,6 +1899,15 @@ where
                     let msgs = self.build_raft_messages(ctx, snap_ctx.msgs);
                     self.send_raft_messages(ctx, msgs);
 
+                    if self.last_compacted_idx == 0
+                        && snap_ctx.last_first_index >= RAFT_INIT_LOG_INDEX
+                    {
+                        // There may be stale logs in raft engine, so schedule a task to clean it
+                        // up. This is a best effort, if TiKV is shutdown before the task is
+                        // handled, there can still be stale logs not being deleted until next
+                        // log gc command is executed. This will delete range [0, last_first_index).
```

Comment:
How next log gc command can delete stale logs? For example, the apply thread executes the log gc command, update and persisted the `truncated_state` but crashes before the log is actually gc. After restart it will use `truncated_state.get_index() + 1` as the `last_compacted_idx` and `first_index`

Reply:
No, after being restarted, `last_compacted_idx` is always initialized to 0.

Reply:
It doesn't help because the `compact_index` of the log gc command is not calculated from `last_compacted_idx` and the apply thread also rejects log gc commands that `compact_index <= first_index`